// @dynamic-labs/client/src/client/client.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var logger = require('@dynamic-labs/logger');
var authModule = require('../modules/authModule/authModule.cjs');
var networksModule = require('../modules/networksModule/networksModule.cjs');
var sdkModule = require('../modules/sdkModule/sdkModule.cjs');
var userInterfaceModule = require('../modules/userInterfaceModule/userInterfaceModule.cjs');
var walletsModule = require('../modules/walletsModule/walletsModule.cjs');
var Extendable = require('../utils/Extendable/Extendable.cjs');
var setupFetchHandler = require('../utils/setupFetchHandler/setupFetchHandler.cjs');
var core = require('./core/core.cjs');

const baseClientExtensionName = 'base';
/**
 * Generates a basic client that allows interacting with Dynamic's SDK, which
 * will be rendered inside a webview.
 *
 * It must be extended with the necessary Extensions that implement the services
 * it will be used for.
 */
const createClient = (props) => {
    logger.Logger.setEnvironmentId(props.environmentId);
    const core$1 = core.createCore(props);
    setupFetchHandler.setupFetchHandler(core$1);
    const baseExtendable = new Extendable.Extendable(core$1);
    return baseExtendable.extend(() => {
        core$1.declaredExtensionNames.push(baseClientExtensionName);
        return {
            /** Module that gives access over authentication state, such as the auth token and user */
            auth: authModule.createAuthModule(core$1),
            /** Module that provides all the networks configured */
            networks: networksModule.createNetworksModule(core$1),
            /** Module that gives insight over the state of the SDK */
            sdk: sdkModule.createSdkModule(core$1),
            /** Module that provide access to UI features */
            ui: userInterfaceModule.createUserInterfaceModule(core$1),
            /** Module that provides access to user's wallets */
            wallets: walletsModule.createWalletsModule(core$1),
        };
    });
};

exports.baseClientExtensionName = baseClientExtensionName;
exports.createClient = createClient;
