// @dynamic-labs/client/src/client/client.js
'use client'
import { Logger } from '@dynamic-labs/logger';
import { createAuthModule } from '../modules/authModule/authModule.js';
import { createNetworksModule } from '../modules/networksModule/networksModule.js';
import { createSdkModule } from '../modules/sdkModule/sdkModule.js';
import { createUserInterfaceModule } from '../modules/userInterfaceModule/userInterfaceModule.js';
import { createWalletsModule } from '../modules/walletsModule/walletsModule.js';
import { Extendable } from '../utils/Extendable/Extendable.js';
import { setupFetchHandler } from '../utils/setupFetchHandler/setupFetchHandler.js';
import { createCore } from './core/core.js';

const baseClientExtensionName = 'base';
/**
 * Generates a basic client that allows interacting with Dynamic's SDK, which
 * will be rendered inside a webview.
 *
 * It must be extended with the necessary Extensions that implement the services
 * it will be used for.
 */
const createClient = (props) => {
    Logger.setEnvironmentId(props.environmentId);
    const core = createCore(props);
    setupFetchHandler(core);
    const baseExtendable = new Extendable(core);
    return baseExtendable.extend(() => {
        core.declaredExtensionNames.push(baseClientExtensionName);
        return {
            /** Module that gives access over authentication state, such as the auth token and user */
            auth: createAuthModule(core),
            /** Module that provides all the networks configured */
            networks: createNetworksModule(core),
            /** Module that gives insight over the state of the SDK */
            sdk: createSdkModule(core),
            /** Module that provide access to UI features */
            ui: createUserInterfaceModule(core),
            /** Module that provides access to user's wallets */
            wallets: createWalletsModule(core),
        };
    });
};

export { baseClientExtensionName, createClient };
