// @zerodev/multi-chain-validator/_cjs/multiChainClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createKernelMultiChainClient = exports.kernelAccountMultiChainClientActions = void 0;
const sdk_1 = require("@zerodev/sdk");
const permissionless_1 = require("permissionless");
const viem_1 = require("viem");
const index_js_1 = require("./actions/index.js");
function kernelAccountMultiChainClientActions({ middleware }) {
    return (client) => ({
        prepareMultiUserOpRequest: (args, validatorType, numOfUserOps, stateOverrides) => (0, index_js_1.prepareMultiUserOpRequest)(client, {
            ...args,
            middleware: {
                ...middleware,
                ...args.middleware
            }
        }, validatorType, numOfUserOps, stateOverrides)
    });
}
exports.kernelAccountMultiChainClientActions = kernelAccountMultiChainClientActions;
const createKernelMultiChainClient = (parameters) => {
    const { key = "Account", name = "Kernel Account Client", bundlerTransport, entryPoint, account } = parameters;
    if (!account) {
        throw new Error("Kernel account is not provided");
    }
    const entryPointVersion = (0, permissionless_1.getEntryPointVersion)(entryPoint);
    const shouldIncludePimlicoProvider = bundlerTransport({}).config.key === "http" &&
        entryPointVersion === "v0.7";
    const client = (0, viem_1.createClient)({
        ...parameters,
        key,
        name,
        transport: (opts) => {
            let _bundlerTransport = bundlerTransport({
                ...opts,
                timeout: bundlerTransport({}).config.timeout,
                retryCount: 0
            });
            if (!shouldIncludePimlicoProvider ||
                (0, sdk_1.isProviderSet)(_bundlerTransport.value?.url, "ALCHEMY"))
                return _bundlerTransport;
            _bundlerTransport = (0, viem_1.http)((0, sdk_1.setPimlicoAsProvider)(_bundlerTransport.value?.url))({
                ...opts,
                timeout: bundlerTransport({}).config.timeout,
                retryCount: 0
            });
            return _bundlerTransport;
        },
        type: "kernelAccountClient"
    });
    let middleware = parameters.middleware;
    if ((!middleware ||
        (typeof middleware !== "function" && !middleware.gasPrice)) &&
        client.transport?.url &&
        (0, sdk_1.isProviderSet)(client.transport.url, "PIMLICO")) {
        const gasPrice = () => (0, sdk_1.getUserOperationGasPrice)(client);
        middleware = {
            ...middleware,
            gasPrice
        };
    }
    return client
        .extend((0, sdk_1.kernelAccountClientActions)({
        middleware
    }))
        .extend(kernelAccountMultiChainClientActions({
        middleware
    }));
};
exports.createKernelMultiChainClient = createKernelMultiChainClient;
