// @zerodev/sdk/_cjs/actions/account-client/invalidateNonce.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.invalidateNonce = void 0;
const smartAccount_1 = require("permissionless/actions/smartAccount");
const utils_1 = require("permissionless/utils");
const utils_2 = require("viem/utils");
const KernelAccountAbi_js_1 = require("../../accounts/kernel/abi/kernel_v_3_0_0/KernelAccountAbi.js");
async function invalidateNonce(client, args) {
    const { account: account_ = client.account, middleware, nonceToSet } = args;
    if (!account_)
        throw new utils_1.AccountOrClientNotFoundError();
    const account = (0, utils_1.parseAccount)(account_);
    return await (0, utils_2.getAction)(client, (smartAccount_1.sendTransaction), "sendTransaction")({
        ...args,
        to: account.address,
        data: (0, utils_2.encodeFunctionData)({
            abi: KernelAccountAbi_js_1.KernelV3AccountAbi,
            functionName: "invalidateNonce",
            args: [nonceToSet]
        }),
        value: 0n,
        account,
        middleware
    });
}
exports.invalidateNonce = invalidateNonce;
