// @dynamic-labs/viem-extension/src/ViemExtension/createPublicClientForDynamic/createPublicClientForDynamic.js
'use client'
import { createPublicClient, custom } from 'viem';
import { createRequestChannel } from '@dynamic-labs/message-transport';

// If we use an arrow function here, typescript complains about non portable type inference
// eslint-disable-next-line prefer-arrow/prefer-arrow-functions
function createPublicClientForDynamic(messageTransport, params) {
    const requestChannel = createRequestChannel(messageTransport);
    const providerProxy = {
        request: (requestArguments) => {
            var _a;
            return requestChannel.request('ethRequest', {
                chainId: params.chain.id,
                method: requestArguments.method,
                params: (_a = requestArguments.params) !== null && _a !== void 0 ? _a : requestArguments.args,
            });
        },
    };
    return createPublicClient(Object.assign(Object.assign({}, params), { transport: custom(providerProxy, {
            key: 'dynamic-client-transport',
            name: 'Dynamic Client Transport',
            retryCount: 3,
            retryDelay: 200,
        }) }));
}

export { createPublicClientForDynamic };
