// @dynamic-labs/utils/src/index.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var assertPackageVersion = require('@dynamic-labs/assert-package-version');
var _package = require('../package.cjs');
var bufferToBase64 = require('./bufferToBase64/bufferToBase64.cjs');
var CancellablePromise = require('./CancellablePromise/CancellablePromise.cjs');
var ceil = require('./ceil/ceil.cjs');
var DeferredPromise = require('./DeferredPromise/DeferredPromise.cjs');
var eip6963Provider = require('./eip6963/eip6963Provider.cjs');
var AccountExistsError = require('./errors/AccountExistsError.cjs');
var ChainalysisError = require('./errors/ChainalysisError.cjs');
var SandboxMaximumThresholdReachedError = require('./errors/SandboxMaximumThresholdReachedError.cjs');
var CustomError = require('./errors/CustomError.cjs');
var DynamicError = require('./errors/DynamicError.cjs');
var EmailAlreadyExistsError = require('./errors/EmailAlreadyExistsError.cjs');
var EmailProviderError = require('./errors/EmailProviderError.cjs');
var EmailVerificationError = require('./errors/EmailVerificationError.cjs');
var GateBlockedError = require('./errors/GateBlockedError.cjs');
var InsufficientFundsError = require('./errors/InsufficientFundsError.cjs');
var MissingEnvironmentIdError = require('./errors/MissingEnvironmentIdError.cjs');
var MissingPublicAddressError = require('./errors/MissingPublicAddressError.cjs');
var NoAccessError = require('./errors/NoAccessError.cjs');
var NotSupportedError = require('./errors/NotSupportedError.cjs');
var SmsVerificationError = require('./errors/SmsVerificationError.cjs');
var SocialAccountAlreadyExistsError = require('./errors/SocialAccountAlreadyExistsError.cjs');
var TransactionGasCannotBeSponsoredError = require('./errors/TransactionGasCannotBeSponsoredError.cjs');
var UserHasAccountWithEmailError = require('./errors/UserHasAccountWithEmailError.cjs');
var UsernameAlreadyExistsError = require('./errors/UsernameAlreadyExistsError.cjs');
var WalletNotDeployedError = require('./errors/WalletNotDeployedError.cjs');
var WalletUsedError = require('./errors/WalletUsedError.cjs');
var UserRejectedTransactionError = require('./errors/UserRejectedTransactionError.cjs');
var TooManyEmailVerificationsError = require('./errors/TooManyEmailVerificationsError.cjs');
var VerificationDataCollectionError = require('./errors/VerificationDataCollectionError.cjs');
var InvalidPhoneNumberError = require('./errors/InvalidPhoneNumberError.cjs');
var MergeAccountsConfirmationError = require('./errors/MergeAccountsConfirmationError.cjs');
var CustomFieldNotUniqueError = require('./errors/CustomFieldNotUniqueError.cjs');
var CookieInvalidError = require('./errors/CookieInvalidError.cjs');
var RequestChannelNotHandledError = require('./errors/RequestChannelNotHandledError.cjs');
var AccessBlockedError = require('./errors/AccessBlockedError.cjs');
var EmbeddedWalletException = require('./errors/EmbeddedWalletException.cjs');
var MfaInvalidOtpError = require('./errors/MfaInvalidOtpError.cjs');
var MfaRateLimitedError = require('./errors/MfaRateLimitedError.cjs');
var ExternalAuthError = require('./errors/ExternalAuthError.cjs');
var WalletAddressMismatchError = require('./errors/WalletAddressMismatchError.cjs');
var UserRejectedRequestError = require('./errors/UserRejectedRequestError.cjs');
var InvalidEmbeddedWalletSessionKeyError = require('./errors/InvalidEmbeddedWalletSessionKeyError.cjs');
var eventTimeline = require('./eventTimeline/eventTimeline.cjs');
var formatNumberText = require('./formatNumberText/formatNumberText.cjs');
var getProvidersFromWindow = require('./getProvidersFromWindow/getProvidersFromWindow.cjs');
var getTLD = require('./getTLD/getTLD.cjs');
var handleMobileWalletRedirect = require('./handleMobileWalletRedirect/handleMobileWalletRedirect.cjs');
var isFunction = require('./isFunction/isFunction.cjs');
var isLedgerAddressViaVerifiedCredentials = require('./isLedgerAddressViaVerifiedCredentials/isLedgerAddressViaVerifiedCredentials.cjs');
var isMobile = require('./isMobile/isMobile.cjs');
var isString = require('./isString/isString.cjs');
var last = require('./last/last.cjs');
var localStorageAsync = require('./localStorageAsync/localStorageAsync.cjs');
var parseChainId = require('./parseChainId/parseChainId.cjs');
var parseIntSafe = require('./parseIntSafe/parseIntSafe.cjs');
var parseNetworks = require('./parseNetworks/parseNetworks.cjs');
var pipe = require('./pipe/pipe.cjs');
var retryableFn = require('./retryableFn/retryableFn.cjs');
var sanitizeName = require('./sanitizeName/sanitizeName.cjs');
var sleep = require('./sleep/sleep.cjs');
var trimEnd = require('./trimEnd/trimEnd.cjs');
var uniq = require('./uniq/uniq.cjs');
var wrapMethodWithCallback = require('./wrapMethodWithCallback/wrapMethodWithCallback.cjs');
var createKeyPair = require('./encryption/createKeyPair/createKeyPair.cjs');
var deriveSharedSecret = require('./encryption/deriveSharedSecret/deriveSharedSecret.cjs');
var convertPublicKeyCryptoKeyToHex = require('./encryption/convertPublicKeyCryptoKeyToHex/convertPublicKeyCryptoKeyToHex.cjs');
var convertPublicKeyHexToCryptoKey = require('./encryption/convertPublicKeyHexToCryptoKey/convertPublicKeyHexToCryptoKey.cjs');
var decryptMessage = require('./encryption/decryptMessage/decryptMessage.cjs');
var encryptMessage = require('./encryption/encryptMessage/encryptMessage.cjs');
var isEncryptedMessage = require('./encryption/isEncryptedMessage/isEncryptedMessage.cjs');
var cloneObjectWithOverrides = require('./cloneObjectWithOverrides/cloneObjectWithOverrides.cjs');
var get = require('./get/get.cjs');
var hexToString = require('./hexToString/hexToString.cjs');
var isHex = require('./isHex/isHex.cjs');
var isInIframe = require('./isInIframe/isInIframe.cjs');
var promiseWithTimeout = require('./promiseWithTimeout/promiseWithTimeout.cjs');
var runSafe = require('./runSafe/runSafe.cjs');
var template = require('./template/template.cjs');
var PlatformService = require('./services/PlatformService/PlatformService.cjs');
var createBrowserPlatformService = require('./services/PlatformService/createBrowserPlatformService/createBrowserPlatformService.cjs');
var FetchService = require('./services/FetchService/FetchService.cjs');
var Oauth2Service = require('./services/Oauth2Service/Oauth2Service.cjs');
var createWindowOauth2Service = require('./services/Oauth2Service/createWindowOauth2Service/createWindowOauth2Service.cjs');
var StorageService = require('./services/StorageService/StorageService.cjs');
var createStorageService = require('./services/StorageService/createStorageService/createStorageService.cjs');
var applyPostfixToStorageService = require('./services/StorageService/applyPostfixToStorageService/applyPostfixToStorageService.cjs');
var connectWithAppleId = require('./services/Oauth2Service/utils/connectWithAppleId/connectWithAppleId.cjs');
var solToLamports = require('./solToLamports/solToLamports.cjs');

assertPackageVersion.assertPackageVersion('@dynamic-labs/utils', _package.version);

exports.bufferToBase64 = bufferToBase64.bufferToBase64;
exports.CancellablePromise = CancellablePromise.CancellablePromise;
exports.ceil = ceil.ceil;
exports.DeferredPromise = DeferredPromise.DeferredPromise;
exports.Eip6963Provider = eip6963Provider.Eip6963Provider;
exports.Eip6963ProviderSingleton = eip6963Provider.Eip6963ProviderSingleton;
exports.isEip9693Event = eip6963Provider.isEip9693Event;
exports.listenToEip6963Events = eip6963Provider.listenToEip6963Events;
exports.AccountExistsError = AccountExistsError.AccountExistsError;
exports.ChainalysisError = ChainalysisError.ChainalysisError;
exports.SandboxMaximumThresholdReachedError = SandboxMaximumThresholdReachedError.SandboxMaximumThresholdReachedError;
exports.CustomError = CustomError.CustomError;
exports.DynamicError = DynamicError.DynamicError;
exports.EmailAlreadyExistsError = EmailAlreadyExistsError.EmailAlreadyExistsError;
exports.EmailProviderError = EmailProviderError.EmailProviderError;
exports.EmailVerificationError = EmailVerificationError.EmailVerificationError;
exports.GateBlockedError = GateBlockedError.GateBlockedError;
exports.InsufficientFundsError = InsufficientFundsError.InsufficientFundsError;
exports.MissingEnvironmentIdError = MissingEnvironmentIdError.MissingEnvironmentIdError;
exports.MissingPublicAddressError = MissingPublicAddressError.MissingPublicAddressError;
exports.NoAccessError = NoAccessError.NoAccessError;
exports.NotSupportedError = NotSupportedError.NotSupportedError;
exports.SmsVerificationError = SmsVerificationError.SmsVerificationError;
exports.SocialAccountAlreadyExistsError = SocialAccountAlreadyExistsError.SocialAccountAlreadyExistsError;
exports.TransactionGasCannotBeSponsoredError = TransactionGasCannotBeSponsoredError.TransactionGasCannotBeSponsoredError;
exports.UserHasAccountWithEmailError = UserHasAccountWithEmailError.UserHasAccountWithEmailError;
exports.UsernameAlreadyExistsError = UsernameAlreadyExistsError.UsernameAlreadyExistsError;
exports.WalletNotDeployedError = WalletNotDeployedError.WalletNotDeployedError;
exports.WalletUsedError = WalletUsedError.WalletUsedError;
exports.UserRejectedTransactionError = UserRejectedTransactionError.UserRejectedTransactionError;
exports.TooManyEmailVerificationsError = TooManyEmailVerificationsError.TooManyEmailVerificationsError;
exports.VerificationDataCollectionError = VerificationDataCollectionError.VerificationDataCollectionError;
exports.InvalidPhoneNumberError = InvalidPhoneNumberError.InvalidPhoneNumberError;
exports.MergeAccountsConfirmationError = MergeAccountsConfirmationError.MergeAccountsConfirmationError;
exports.CustomFieldNotUniqueError = CustomFieldNotUniqueError.CustomFieldNotUniqueError;
exports.CookieInvalidError = CookieInvalidError.CookieInvalidError;
exports.RequestChannelNotHandledError = RequestChannelNotHandledError.RequestChannelNotHandledError;
exports.AccessBlockedError = AccessBlockedError.AccessBlockedError;
exports.EmbeddedWalletException = EmbeddedWalletException.EmbeddedWalletException;
exports.MfaInvalidOtpError = MfaInvalidOtpError.MfaInvalidOtpError;
exports.MfaRateLimitedError = MfaRateLimitedError.MfaRateLimitedError;
exports.ExternalAuthError = ExternalAuthError.ExternalAuthError;
exports.WalletAddressMismatchError = WalletAddressMismatchError.WalletAddressMismatchError;
exports.UserRejectedRequestError = UserRejectedRequestError.UserRejectedRequestError;
exports.InvalidEmbeddedWalletSessionKeyError = InvalidEmbeddedWalletSessionKeyError.InvalidEmbeddedWalletSessionKeyError;
exports.createEventTimeline = eventTimeline.createEventTimeline;
exports.formatNumberText = formatNumberText.formatNumberText;
exports.getProvidersFromWindow = getProvidersFromWindow.getProvidersFromWindow;
exports.getTLD = getTLD.getTLD;
exports.handleMobileWalletRedirect = handleMobileWalletRedirect.handleMobileWalletRedirect;
exports.isFunction = isFunction.isFunction;
exports.isLedgerAddressViaVerifiedCredentials = isLedgerAddressViaVerifiedCredentials.isLedgerAddressViaVerifiedCredentials;
exports.getAndroidVersion = isMobile.getAndroidVersion;
exports.isAndroid = isMobile.isAndroid;
exports.isIOS = isMobile.isIOS;
exports.isIPad = isMobile.isIPad;
exports.isIPhone = isMobile.isIPhone;
exports.isIPhone8OrEarlier = isMobile.isIPhone8OrEarlier;
exports.isLegacySafari = isMobile.isLegacySafari;
exports.isMobile = isMobile.isMobile;
exports.isSafariBrowser = isMobile.isSafariBrowser;
exports.isSamsungBrowser = isMobile.isSamsungBrowser;
exports.isWindows = isMobile.isWindows;
exports.isString = isString.isString;
exports.last = last.last;
exports.getItemAsync = localStorageAsync.getItemAsync;
exports.removeItemAsync = localStorageAsync.removeItemAsync;
exports.setItemAsync = localStorageAsync.setItemAsync;
exports.parseChainId = parseChainId.parseChainId;
exports.parseIntSafe = parseIntSafe.parseIntSafe;
exports.parseCosmosNetworks = parseNetworks.parseCosmosNetworks;
exports.parseEvmNetworks = parseNetworks.parseEvmNetworks;
exports.pipe = pipe.pipe;
exports.FALLBACK_UNDEFINED = retryableFn.FALLBACK_UNDEFINED;
exports.retryableFn = retryableFn.retryableFn;
exports.sanitizeName = sanitizeName.sanitizeName;
exports.sleep = sleep.sleep;
exports.trimEnd = trimEnd.trimEnd;
exports.uniq = uniq.uniq;
exports.wrapMethodWithCallback = wrapMethodWithCallback.wrapMethodWithCallback;
exports.createKeyPair = createKeyPair.createKeyPair;
exports.deriveSharedSecret = deriveSharedSecret.deriveSharedSecret;
exports.convertPublicKeyCryptoKeyToHex = convertPublicKeyCryptoKeyToHex.convertPublicKeyCryptoKeyToHex;
exports.convertPublicKeyHexToCryptoKey = convertPublicKeyHexToCryptoKey.convertPublicKeyHexToCryptoKey;
exports.decryptMessage = decryptMessage.decryptMessage;
exports.encryptMessage = encryptMessage.encryptMessage;
exports.isEncryptedMessage = isEncryptedMessage.isEncryptedMessage;
exports.cloneObjectWithOverrides = cloneObjectWithOverrides.cloneObjectWithOverrides;
exports.get = get.get;
exports.hexToString = hexToString.hexToString;
exports.isHex = isHex.isHex;
exports.isInIframe = isInIframe.isInIframe;
exports.promiseWithTimeout = promiseWithTimeout.promiseWithTimeout;
exports.runSafe = runSafe.runSafe;
exports.template = template.template;
exports.PlatformService = PlatformService.PlatformService;
exports.createBrowserPlatformService = createBrowserPlatformService.createBrowserPlatformService;
exports.FetchService = FetchService.FetchService;
exports.Oauth2Service = Oauth2Service.Oauth2Service;
exports.createWindowOauth2Service = createWindowOauth2Service.createWindowOauth2Service;
exports.StorageService = StorageService.StorageService;
exports.createStorageService = createStorageService.createStorageService;
exports.applyPostfixToStorageService = applyPostfixToStorageService.applyPostfixToStorageService;
exports.connectWithAppleId = connectWithAppleId.connectWithAppleId;
exports.solToLamports = solToLamports.solToLamports;
