// @dynamic-labs/sdk-api-core/src/models/SolanaChainType.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.SolanaChainType = void 0;
(function (SolanaChainType) {
    SolanaChainType["Mainnet"] = "mainnet";
    SolanaChainType["Devnet"] = "devnet";
})(exports.SolanaChainType || (exports.SolanaChainType = {}));
function SolanaChainTypeFromJSON(json) {
    return SolanaChainTypeFromJSONTyped(json);
}
function SolanaChainTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function SolanaChainTypeToJSON(value) {
    return value;
}

exports.SolanaChainTypeFromJSON = SolanaChainTypeFromJSON;
exports.SolanaChainTypeFromJSONTyped = SolanaChainTypeFromJSONTyped;
exports.SolanaChainTypeToJSON = SolanaChainTypeToJSON;
