// @zerodev/sdk/_esm/actions/account-client/changeSudoValidator.js
import { sendUserOperation } from "permissionless/actions/smartAccount";
import { AccountOrClientNotFoundError, parseAccount } from "permissionless/utils";
import { zeroAddress } from "viem";
import { concatHex, encodeFunctionData, getAction, pad } from "viem/utils";
import { KernelV3_1AccountAbi } from "../../accounts/kernel/abi/kernel_v_3_1/KernelAccountAbi.js";
import { KERNEL_V3_0, KERNEL_V3_1, KernelVersionToAddressesMap, VALIDATOR_TYPE } from "../../constants.js";
export async function changeSudoValidator(client, args) {
    const { account: account_ = client.account, sudoValidator, hook } = args;
    if (!account_)
        throw new AccountOrClientNotFoundError();
    const account = parseAccount(account_);
    let rootValidatorId;
    if ([VALIDATOR_TYPE.PERMISSION, VALIDATOR_TYPE.SECONDARY].includes(VALIDATOR_TYPE[sudoValidator.validatorType])) {
        rootValidatorId = concatHex([
            VALIDATOR_TYPE[sudoValidator.validatorType],
            pad(sudoValidator.getIdentifier(), {
                size: 20,
                dir: "right"
            })
        ]);
    }
    else {
        throw new Error(`Cannot change sudo validator to type ${sudoValidator.validatorType}`);
    }
    const validatorData = await sudoValidator.getEnableData(account.address);
    const hookId = hook?.getIdentifier() ?? zeroAddress;
    const hookData = (await hook?.getEnableData(account.address)) ?? "0x";
    /**
     * @dev Kernel v3.0 does not support changeRootValidator directly, so we need to use delegatecall to call changeRootValidator on Kernel v3.1 implementation contract
     */
    if (account.kernelVersion === KERNEL_V3_0) {
        return await getAction(client, (sendUserOperation), "sendUserOperation")({
            ...args,
            userOperation: {
                callData: await account.encodeCallData({
                    to: KernelVersionToAddressesMap[KERNEL_V3_1]
                        .accountImplementationAddress,
                    value: 0n,
                    data: encodeFunctionData({
                        abi: KernelV3_1AccountAbi,
                        functionName: "changeRootValidator",
                        args: [rootValidatorId, hookId, validatorData, hookData]
                    }),
                    callType: "delegatecall"
                })
            }
        });
    }
    /**
     * @dev Kernel v3.1 supports changeRootValidator directly
     */
    return await getAction(client, (sendUserOperation), "sendUserOperation")({
        ...args,
        userOperation: {
            callData: await account.encodeCallData({
                to: account.address,
                value: 0n,
                data: encodeFunctionData({
                    abi: KernelV3_1AccountAbi,
                    functionName: "changeRootValidator",
                    args: [rootValidatorId, hookId, validatorData, hookData]
                })
            })
        }
    });
}
