// @zerodev/sdk/_esm/actions/account-client/getUserOperationGasPrice.js
import { getBlock, getChainId } from "viem/actions";
import { arbitrum } from "viem/chains";
import { isProviderSet } from "../../clients/utils.js";
export const getUserOperationGasPrice = async (client) => {
    if (client.transport?.url &&
        (isProviderSet(client.transport.url, "ALCHEMY") ||
            isProviderSet(client.transport.url, "ZERODEV") ||
            isProviderSet(client.transport.url, "CONDUIT"))) {
        const [fee, block] = await Promise.all([
            client.request({
                // @ts-ignore
                method: "rundler_maxPriorityFeePerGas",
                params: []
                // biome-ignore lint/suspicious/noExplicitAny: <explanation>
            }),
            getBlock(client, { blockTag: "latest" })
        ]);
        const chainId = client.chain
            ? client.chain.id
            : await getChainId(client);
        const baseFeeMultiplier = chainId === arbitrum.id ? 1.05 : 1.5;
        const priorityFeeMultiplier = chainId === arbitrum.id ? 1.0 : 1.25;
        const denominator = 100;
        const multiply = (base, multiplier) => (base * BigInt(multiplier * denominator)) / BigInt(denominator);
        const maxPriorityFeePerGas = multiply(BigInt(fee), priorityFeeMultiplier);
        const baseFeePerGas = multiply(block.baseFeePerGas, baseFeeMultiplier);
        const maxFeePerGas = baseFeePerGas + maxPriorityFeePerGas;
        return { maxFeePerGas, maxPriorityFeePerGas };
    }
    const gasPrice = await client.request({
        method: "zd_getUserOperationGasPrice",
        params: []
    });
    return {
        maxFeePerGas: BigInt(gasPrice.standard.maxFeePerGas),
        maxPriorityFeePerGas: BigInt(gasPrice.standard.maxPriorityFeePerGas)
    };
};
