// @zerodev/sdk/_esm/actions/account-client/uninstallPlugin.js
import { sendTransactions } from "permissionless/actions/smartAccount";
import { AccountOrClientNotFoundError, parseAccount } from "permissionless/utils";
import { concatHex, encodeFunctionData, getAction, pad } from "viem/utils";
import { KernelV3AccountAbi } from "../../accounts/kernel/abi/kernel_v_3_0_0/KernelAccountAbi.js";
import { VALIDATOR_TYPE } from "../../constants.js";
export async function uninstallPlugin(client, args) {
    const { account: account_ = client.account, middleware, plugin, hook } = args;
    if (!account_)
        throw new AccountOrClientNotFoundError();
    const account = parseAccount(account_);
    let validatorId;
    if ([VALIDATOR_TYPE.PERMISSION, VALIDATOR_TYPE.SECONDARY].includes(VALIDATOR_TYPE[plugin.validatorType])) {
        validatorId = concatHex([
            VALIDATOR_TYPE[plugin.validatorType],
            pad(plugin.getIdentifier(), {
                size: 20,
                dir: "right"
            })
        ]);
    }
    else {
        throw new Error(`Cannot uninstall ${plugin.validatorType} plugin`);
    }
    const validatorData = await plugin.getEnableData(account.address);
    const hookData = (await hook?.getEnableData(account.address)) ?? "0x";
    return await getAction(client, (sendTransactions), "sendTransactions")({
        ...args,
        transactions: [
            {
                to: account.address,
                data: encodeFunctionData({
                    abi: KernelV3AccountAbi,
                    functionName: "uninstallValidation",
                    args: [validatorId, validatorData, hookData]
                }),
                value: 0n
            }
        ],
        account,
        middleware
    });
}
