// @zerodev/ecdsa-validator/_esm/toECDSAValidatorPlugin.js
import { validateKernelVersionWithEntryPoint } from "@zerodev/sdk";
import { satisfiesRange } from "@zerodev/sdk";
import { getUserOperationHash } from "permissionless";
import { SignTransactionNotSupportedBySmartAccount } from "permissionless/accounts";
import { zeroAddress } from "viem";
import { toAccount } from "viem/accounts";
import { signMessage, signTypedData } from "viem/actions";
import { getChainId } from "viem/actions";
import { kernelVersionRangeToValidator } from "./constants.js";
export const getValidatorAddress = (entryPointAddress, kernelVersion, validatorAddress) => {
    validateKernelVersionWithEntryPoint(entryPointAddress, kernelVersion);
    const ecdsaValidatorAddress = Object.entries(kernelVersionRangeToValidator).find(([range]) => satisfiesRange(kernelVersion, range))?.[1];
    if (!ecdsaValidatorAddress && !validatorAddress) {
        throw new Error(`Validator not found for Kernel version: ${kernelVersion}`);
    }
    return validatorAddress ?? ecdsaValidatorAddress ?? zeroAddress;
};
export async function signerToEcdsaValidator(client, { signer, entryPoint: entryPointAddress, kernelVersion, validatorAddress: _validatorAddress }) {
    const validatorAddress = getValidatorAddress(entryPointAddress, kernelVersion, _validatorAddress);
    // Get the private key related account
    const viemSigner = {
        ...signer,
        signTransaction: (_, __) => {
            throw new SignTransactionNotSupportedBySmartAccount();
        }
    };
    // Fetch chain id
    const chainId = await getChainId(client);
    // Build the EOA Signer
    const account = toAccount({
        address: viemSigner.address,
        async signMessage({ message }) {
            return signMessage(client, { account: viemSigner, message });
        },
        async signTransaction(_, __) {
            throw new SignTransactionNotSupportedBySmartAccount();
        },
        async signTypedData(typedData) {
            return signTypedData(client, {
                account: viemSigner,
                ...typedData
            });
        }
    });
    return {
        ...account,
        supportedKernelVersions: kernelVersion,
        validatorType: "SECONDARY",
        address: validatorAddress,
        source: "ECDSAValidator",
        getIdentifier() {
            return validatorAddress;
        },
        async getEnableData() {
            return viemSigner.address;
        },
        async getNonceKey(_accountAddress, customNonceKey) {
            if (customNonceKey) {
                return customNonceKey;
            }
            return 0n;
        },
        // Sign a user operation
        async signUserOperation(userOperation) {
            const hash = getUserOperationHash({
                userOperation: {
                    ...userOperation,
                    signature: "0x"
                },
                entryPoint: entryPointAddress,
                chainId: chainId
            });
            const signature = await signMessage(client, {
                account: viemSigner,
                message: { raw: hash }
            });
            return signature;
        },
        // Get simple dummy signature
        async getDummySignature() {
            return "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c";
        },
        async isEnabled(_kernelAccountAddress, _selector) {
            return false;
        }
    };
}
