// @zerodev/multi-chain-validator/_esm/multiChainClient.js
import { getUserOperationGasPrice, isProviderSet, kernelAccountClientActions, setPimlicoAsProvider } from "@zerodev/sdk";
import { getEntryPointVersion } from "permissionless";
import { http, createClient } from "viem";
import { prepareMultiUserOpRequest } from "./actions/index.js";
export function kernelAccountMultiChainClientActions({ middleware }) {
    return (client) => ({
        prepareMultiUserOpRequest: (args, validatorType, numOfUserOps, stateOverrides) => prepareMultiUserOpRequest(client, {
            ...args,
            middleware: {
                ...middleware,
                ...args.middleware
            }
        }, validatorType, numOfUserOps, stateOverrides)
    });
}
export const createKernelMultiChainClient = (parameters) => {
    const { key = "Account", name = "Kernel Account Client", bundlerTransport, entryPoint, account } = parameters;
    if (!account) {
        throw new Error("Kernel account is not provided");
    }
    const entryPointVersion = getEntryPointVersion(entryPoint);
    const shouldIncludePimlicoProvider = bundlerTransport({}).config.key === "http" &&
        entryPointVersion === "v0.7";
    const client = createClient({
        ...parameters,
        key,
        name,
        transport: (opts) => {
            let _bundlerTransport = bundlerTransport({
                ...opts,
                timeout: bundlerTransport({}).config.timeout,
                retryCount: 0
            });
            if (!shouldIncludePimlicoProvider ||
                isProviderSet(_bundlerTransport.value?.url, "ALCHEMY"))
                return _bundlerTransport;
            _bundlerTransport = http(setPimlicoAsProvider(_bundlerTransport.value?.url))({
                ...opts,
                timeout: bundlerTransport({}).config.timeout,
                retryCount: 0
            });
            return _bundlerTransport;
        },
        type: "kernelAccountClient"
    });
    let middleware = parameters.middleware;
    if ((!middleware ||
        (typeof middleware !== "function" && !middleware.gasPrice)) &&
        client.transport?.url &&
        isProviderSet(client.transport.url, "PIMLICO")) {
        const gasPrice = () => getUserOperationGasPrice(client);
        middleware = {
            ...middleware,
            gasPrice
        };
    }
    return client
        .extend(kernelAccountClientActions({
        middleware
    }))
        .extend(kernelAccountMultiChainClientActions({
        middleware
    }));
};
