// @zerodev/sdk/_cjs/clients/decorators/kernel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.kernelAccountClientActions = exports.zerodevPaymasterActions = void 0;
const permissionless_1 = require("permissionless");
const getKernelV3ModuleCurrentNonce_js_1 = require("../../actions/account-client/getKernelV3ModuleCurrentNonce.js");
const getUserOperationGasPrice_js_1 = require("../../actions/account-client/getUserOperationGasPrice.js");
const invalidateNonce_js_1 = require("../../actions/account-client/invalidateNonce.js");
const index_js_1 = require("../../actions/index.js");
const estimateGasInERC20_js_1 = require("../../actions/paymaster/estimateGasInERC20.js");
const sponsorUserOperation_js_1 = require("../../actions/paymaster/sponsorUserOperation.js");
const zerodevPaymasterActions = (entryPointAddress) => (client) => ({
    sponsorUserOperation: async (args) => (0, sponsorUserOperation_js_1.sponsorUserOperation)(client, {
        ...args,
        entryPoint: entryPointAddress
    }),
    estimateGasInERC20: async (args) => (0, estimateGasInERC20_js_1.estimateGasInERC20)(client, args),
    sponsorUserOperationEip7677: async (args, bundlerClient, stateOverrides) => (0, index_js_1.sponsorUserOperationEip7677)(client, args, bundlerClient, stateOverrides)
});
exports.zerodevPaymasterActions = zerodevPaymasterActions;
function kernelAccountClientActions({ middleware }) {
    return (client) => ({
        ...(0, permissionless_1.smartAccountActions)({ middleware })(client),
        signUserOperation: (args) => (0, index_js_1.signUserOperation)(client, {
            ...args,
            middleware
        }),
        getUserOperationGasPrice: async () => (0, getUserOperationGasPrice_js_1.getUserOperationGasPrice)(client),
        uninstallPlugin: async (args) => (0, index_js_1.uninstallPlugin)(client, {
            ...args,
            middleware
        }),
        changeSudoValidator: async (args) => (0, index_js_1.changeSudoValidator)(client, {
            ...args,
            middleware
        }),
        invalidateNonce: async (args) => (0, invalidateNonce_js_1.invalidateNonce)(client, {
            ...args,
            middleware
        }),
        getKernelV3ModuleCurrentNonce: async (args) => (0, getKernelV3ModuleCurrentNonce_js_1.getKernelV3ModuleCurrentNonce)(client, {
            ...args,
            middleware
        })
    });
}
exports.kernelAccountClientActions = kernelAccountClientActions;
