// @zerodev/sdk/_cjs/actions/paymaster/sponsorUserOperationEip7677.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sponsorUserOperationEip7677 = void 0;
const permissionless_1 = require("permissionless");
const experimental_1 = require("permissionless/experimental");
const sponsorUserOperationEip7677 = async (client, args, bundlerClient, stateOverrides) => {
    const { entryPoint: entryPointAddress, userOperation } = args;
    const chain = client.chain;
    const paymasterClient = client.extend((0, experimental_1.paymasterActionsEip7677)(entryPointAddress));
    const stubData = await paymasterClient.getPaymasterStubData({
        userOperation: userOperation,
        chain
    });
    const stubUserOperation = {
        ...userOperation,
        ...stubData
    };
    const gas = (await bundlerClient.estimateUserOperationGas({
        userOperation: stubUserOperation
    }, stateOverrides));
    const userOperationWithGas = {
        ...stubUserOperation,
        callGasLimit: gas.callGasLimit,
        verificationGasLimit: gas.verificationGasLimit,
        preVerificationGas: gas.preVerificationGas
    };
    const paymasterData = await paymasterClient.getPaymasterData({
        userOperation: userOperationWithGas,
        chain
    });
    if (entryPointAddress === permissionless_1.ENTRYPOINT_ADDRESS_V06) {
        const paymasterDataV06 = paymasterData;
        return {
            callGasLimit: BigInt(gas.callGasLimit),
            verificationGasLimit: BigInt(gas.verificationGasLimit),
            preVerificationGas: BigInt(gas.preVerificationGas),
            paymasterAndData: paymasterDataV06?.paymasterAndData,
            maxFeePerGas: BigInt(userOperation.maxFeePerGas),
            maxPriorityFeePerGas: BigInt(userOperation.maxPriorityFeePerGas)
        };
    }
    const stubDataV07 = stubData;
    const paymasterDataV07 = paymasterData;
    return {
        callGasLimit: BigInt(gas.callGasLimit),
        verificationGasLimit: BigInt(gas.verificationGasLimit),
        preVerificationGas: BigInt(gas.preVerificationGas),
        paymaster: paymasterDataV07.paymaster,
        paymasterData: paymasterDataV07.paymasterData,
        paymasterVerificationGasLimit: stubDataV07.paymasterVerificationGasLimit &&
            BigInt(stubDataV07.paymasterVerificationGasLimit),
        paymasterPostOpGasLimit: stubDataV07?.paymasterPostOpGasLimit &&
            BigInt(stubDataV07.paymasterPostOpGasLimit),
        maxFeePerGas: BigInt(userOperation.maxFeePerGas),
        maxPriorityFeePerGas: BigInt(userOperation.maxPriorityFeePerGas)
    };
};
exports.sponsorUserOperationEip7677 = sponsorUserOperationEip7677;
