// @dynamic-labs/client/src/client/core/modules/manifestModule/validateManifest/validateManifest.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/**
 * Maps each required field of the client manifest to an error message that explains
 * what the customer can do if his app runs into the issue where this field is missing from his manifest.
 */
const mandatoryFields = {
    clientVersion: "We were unable to retrieve the Dynamic Client's version from your app. Please contact us about this.",
    environmentId: 'You must provide an environment id to the Dynamic Client.',
    platform: 'Missing platform specific extension, like WebExtension or ReactNativeExtension. See how to setup Dynamic Client in our docs for help.',
    redirectUrl: 'Missing platform specific extension, like WebExtension or ReactNativeExtension. See how to setup Dynamic Client in our docs for help.',
};
const validateManifest = (partialManifest) => {
    /**
     * If any of these fields are missing, we show the user an error message.
     * You should map to each required field a description of what the user can do to fix this issue.
     *
     * And if you add a new required field to DynamicClient, TS will automatically force you to
     * also include a description here :)
     */
    for (const [field, description] of Object.entries(mandatoryFields))
        if (partialManifest[field] === undefined)
            throw new Error(`Missing client manifest field "${field}": ${description}`);
    return partialManifest;
};

exports.mandatoryFields = mandatoryFields;
exports.validateManifest = validateManifest;
