// @zerodev/sdk/_esm/clients/decorators/kernel.js
import { smartAccountActions } from "permissionless";
import { getKernelV3ModuleCurrentNonce } from "../../actions/account-client/getKernelV3ModuleCurrentNonce.js";
import { getUserOperationGasPrice } from "../../actions/account-client/getUserOperationGasPrice.js";
import { invalidateNonce } from "../../actions/account-client/invalidateNonce.js";
import { changeSudoValidator, signUserOperation, sponsorUserOperationEip7677, uninstallPlugin } from "../../actions/index.js";
import { estimateGasInERC20 } from "../../actions/paymaster/estimateGasInERC20.js";
import { sponsorUserOperation } from "../../actions/paymaster/sponsorUserOperation.js";
export const zerodevPaymasterActions = (entryPointAddress) => (client) => ({
    sponsorUserOperation: async (args) => sponsorUserOperation(client, {
        ...args,
        entryPoint: entryPointAddress
    }),
    estimateGasInERC20: async (args) => estimateGasInERC20(client, args),
    sponsorUserOperationEip7677: async (args, bundlerClient, stateOverrides) => sponsorUserOperationEip7677(client, args, bundlerClient, stateOverrides)
});
export function kernelAccountClientActions({ middleware }) {
    return (client) => ({
        ...smartAccountActions({ middleware })(client),
        signUserOperation: (args) => signUserOperation(client, {
            ...args,
            middleware
        }),
        getUserOperationGasPrice: async () => getUserOperationGasPrice(client),
        uninstallPlugin: async (args) => uninstallPlugin(client, {
            ...args,
            middleware
        }),
        changeSudoValidator: async (args) => changeSudoValidator(client, {
            ...args,
            middleware
        }),
        invalidateNonce: async (args) => invalidateNonce(client, {
            ...args,
            middleware
        }),
        getKernelV3ModuleCurrentNonce: async (args) => getKernelV3ModuleCurrentNonce(client, {
            ...args,
            middleware
        })
    });
}
