// @dynamic-labs/client/src/modules/userInterfaceModule/userInterfaceModule.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var messageTransport = require('@dynamic-labs/message-transport');
var pickListenerActions = require('../../utils/pickListenerActions/pickListenerActions.cjs');

const createUserInterfaceModule = (core) => {
    const userInterfaceRequestChannel = messageTransport.createRequestChannel(core.messageTransport);
    const messageEvents = messageTransport.createEventEmitterForMessages({
        eventNames: messageTransport.userInterfaceEventNames,
        messageTransport: core.messageTransport,
    });
    return Object.assign(pickListenerActions.pickListenerActions(messageEvents), {
        auth: {
            hide: () => userInterfaceRequestChannel.emit('hideAuthFlow'),
            show: () => userInterfaceRequestChannel.emit('openAuthFlow'),
        },
        userProfile: {
            hide: () => userInterfaceRequestChannel.emit('hideUserProfile'),
            show: () => userInterfaceRequestChannel.emit('openUserProfile'),
        },
        wallets: {
            revealEmbeddedWalletKey: (params) => userInterfaceRequestChannel.emit('revealEmbeddedWalletKey', params),
        },
    });
};

exports.createUserInterfaceModule = createUserInterfaceModule;
