// @zerodev/multi-chain-validator/_cjs/actions/prepareMultiUserOpRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareMultiUserOpRequest = void 0;
const permissionless_1 = require("permissionless");
const actions_1 = require("viem/actions");
const utils_1 = require("viem/utils");
const ecdsaGetMultiUserOpDummySignature_js_1 = require("../ecdsa/ecdsaGetMultiUserOpDummySignature.js");
const webauthnGetMultiUserOpDummySignature_js_1 = require("../webauthn/webauthnGetMultiUserOpDummySignature.js");
const type_js_1 = require("./type.js");
async function prepareMultiUserOpRequest(client, args, validatorType, numOfUserOps, stateOverrides) {
    const { account: account_ = client.account, userOperation: partialUserOperation, middleware } = args;
    if (!account_)
        throw new permissionless_1.AccountOrClientNotFoundError();
    const account = (0, permissionless_1.parseAccount)(account_);
    const [sender, nonce, factory, factoryData, callData, gasEstimation] = await Promise.all([
        partialUserOperation.sender || account.address,
        partialUserOperation.nonce || account.getNonce(),
        partialUserOperation.factory || account.getFactory(),
        partialUserOperation.factoryData || account.getFactoryData(),
        partialUserOperation.callData,
        !partialUserOperation.maxFeePerGas ||
            !partialUserOperation.maxPriorityFeePerGas
            ? (0, actions_1.estimateFeesPerGas)(account.client)
            : undefined
    ]);
    const userOperation = {
        sender,
        nonce,
        factory: factory,
        factoryData: factoryData,
        callData,
        callGasLimit: partialUserOperation.callGasLimit || BigInt(0),
        verificationGasLimit: partialUserOperation.verificationGasLimit || BigInt(0),
        preVerificationGas: partialUserOperation.preVerificationGas || BigInt(0),
        maxFeePerGas: partialUserOperation.maxFeePerGas ||
            gasEstimation?.maxFeePerGas ||
            BigInt(0),
        maxPriorityFeePerGas: partialUserOperation.maxPriorityFeePerGas ||
            gasEstimation?.maxPriorityFeePerGas ||
            BigInt(0),
        signature: partialUserOperation.signature || "0x"
    };
    if (typeof middleware === "function") {
        return middleware({
            userOperation,
            entryPoint: account.entryPoint
        });
    }
    if (middleware && typeof middleware !== "function" && middleware.gasPrice) {
        const gasPrice = await middleware.gasPrice();
        userOperation.maxFeePerGas = gasPrice.maxFeePerGas;
        userOperation.maxPriorityFeePerGas = gasPrice.maxPriorityFeePerGas;
    }
    if (!userOperation.maxFeePerGas || !userOperation.maxPriorityFeePerGas) {
        const estimateGas = await (0, actions_1.estimateFeesPerGas)(account.client);
        userOperation.maxFeePerGas =
            userOperation.maxFeePerGas || estimateGas.maxFeePerGas;
        userOperation.maxPriorityFeePerGas =
            userOperation.maxPriorityFeePerGas ||
                estimateGas.maxPriorityFeePerGas;
    }
    const chainId = await (0, actions_1.getChainId)(client);
    if (userOperation.signature === "0x") {
        if (validatorType === type_js_1.ValidatorType.ECDSA) {
            userOperation.signature = await (0, ecdsaGetMultiUserOpDummySignature_js_1.ecdsaGetMultiUserOpDummySignature)(userOperation, numOfUserOps, account.entryPoint, chainId);
        }
        if (validatorType === type_js_1.ValidatorType.WEBAUTHN) {
            userOperation.signature =
                await (0, webauthnGetMultiUserOpDummySignature_js_1.webauthnGetMultiUserOpDummySignature)(userOperation, numOfUserOps, account.entryPoint, chainId);
        }
    }
    if (middleware &&
        typeof middleware !== "function" &&
        middleware.sponsorUserOperation) {
        const sponsorUserOperationData = (await middleware.sponsorUserOperation({
            userOperation,
            entryPoint: account.entryPoint
        }));
        userOperation.callGasLimit = sponsorUserOperationData.callGasLimit;
        userOperation.verificationGasLimit =
            sponsorUserOperationData.verificationGasLimit;
        userOperation.preVerificationGas =
            sponsorUserOperationData.preVerificationGas;
        userOperation.paymaster = sponsorUserOperationData.paymaster;
        userOperation.paymasterVerificationGasLimit =
            sponsorUserOperationData.paymasterVerificationGasLimit;
        userOperation.paymasterPostOpGasLimit =
            sponsorUserOperationData.paymasterPostOpGasLimit;
        userOperation.paymasterData = sponsorUserOperationData.paymasterData;
        userOperation.maxFeePerGas =
            sponsorUserOperationData.maxFeePerGas || userOperation.maxFeePerGas;
        userOperation.maxPriorityFeePerGas =
            sponsorUserOperationData.maxPriorityFeePerGas ||
                userOperation.maxPriorityFeePerGas;
        return userOperation;
    }
    if (!userOperation.callGasLimit ||
        !userOperation.verificationGasLimit ||
        !userOperation.preVerificationGas) {
        const gasParameters = await (0, utils_1.getAction)(client, (permissionless_1.estimateUserOperationGas), "estimateUserOperationGas")({
            userOperation,
            entryPoint: account.entryPoint
        }, stateOverrides);
        userOperation.callGasLimit |= gasParameters.callGasLimit;
        userOperation.verificationGasLimit =
            userOperation.verificationGasLimit ||
                gasParameters.verificationGasLimit;
        userOperation.preVerificationGas =
            userOperation.preVerificationGas || gasParameters.preVerificationGas;
        userOperation.paymasterPostOpGasLimit =
            userOperation.paymasterPostOpGasLimit ||
                gasParameters.paymasterPostOpGasLimit;
        userOperation.paymasterPostOpGasLimit =
            userOperation.paymasterPostOpGasLimit ||
                gasParameters.paymasterPostOpGasLimit;
    }
    return userOperation;
}
exports.prepareMultiUserOpRequest = prepareMultiUserOpRequest;
