// @zerodev/multi-chain-validator/_esm/ecdsa/ecdsaSignUserOps.js
import { MerkleTree } from "merkletreejs";
import { getUserOperationHash } from "permissionless";
import { concatHex, encodeAbiParameters, keccak256 } from "viem";
export async function ecdsaSignUserOps({ account, multiUserOps, entryPoint: entryPointAddress }) {
    const userOpHashes = multiUserOps.map((multiUserOp, _index) => {
        return getUserOperationHash({
            userOperation: {
                ...multiUserOp.userOperation,
                signature: "0x"
            },
            entryPoint: entryPointAddress,
            chainId: multiUserOp.chainId
        });
    });
    const merkleTree = new MerkleTree(userOpHashes, keccak256, {
        sortPairs: true
    });
    const merkleRoot = merkleTree.getHexRoot();
    const ecdsaSig = await account.kernelPluginManager.signMessage({
        message: {
            raw: merkleRoot
        }
    });
    const merkleProofs = userOpHashes.map((hash) => {
        return merkleTree.getHexProof(hash);
    });
    const multiChainSigs = multiUserOps.map((_, index) => {
        const merkleProof = merkleProofs[index];
        const encodedMerkleProof = encodeAbiParameters([{ name: "proof", type: "bytes32[]" }], [merkleProof]);
        return concatHex([ecdsaSig, merkleRoot, encodedMerkleProof]);
    });
    const signedMultiUserOps = multiUserOps.map((multiUserOp, index) => {
        return {
            ...multiUserOp.userOperation,
            signature: multiChainSigs[index]
        };
    });
    return signedMultiUserOps;
}
