// @zerodev/ecdsa-validator/_cjs/toECDSAValidatorPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.signerToEcdsaValidator = exports.getValidatorAddress = void 0;
const sdk_1 = require("@zerodev/sdk");
const sdk_2 = require("@zerodev/sdk");
const permissionless_1 = require("permissionless");
const accounts_1 = require("permissionless/accounts");
const viem_1 = require("viem");
const accounts_2 = require("viem/accounts");
const actions_1 = require("viem/actions");
const actions_2 = require("viem/actions");
const constants_js_1 = require("./constants.js");
const getValidatorAddress = (entryPointAddress, kernelVersion, validatorAddress) => {
    (0, sdk_1.validateKernelVersionWithEntryPoint)(entryPointAddress, kernelVersion);
    const ecdsaValidatorAddress = Object.entries(constants_js_1.kernelVersionRangeToValidator).find(([range]) => (0, sdk_2.satisfiesRange)(kernelVersion, range))?.[1];
    if (!ecdsaValidatorAddress && !validatorAddress) {
        throw new Error(`Validator not found for Kernel version: ${kernelVersion}`);
    }
    return validatorAddress ?? ecdsaValidatorAddress ?? viem_1.zeroAddress;
};
exports.getValidatorAddress = getValidatorAddress;
async function signerToEcdsaValidator(client, { signer, entryPoint: entryPointAddress, kernelVersion, validatorAddress: _validatorAddress }) {
    const validatorAddress = (0, exports.getValidatorAddress)(entryPointAddress, kernelVersion, _validatorAddress);
    const viemSigner = {
        ...signer,
        signTransaction: (_, __) => {
            throw new accounts_1.SignTransactionNotSupportedBySmartAccount();
        }
    };
    const chainId = await (0, actions_2.getChainId)(client);
    const account = (0, accounts_2.toAccount)({
        address: viemSigner.address,
        async signMessage({ message }) {
            return (0, actions_1.signMessage)(client, { account: viemSigner, message });
        },
        async signTransaction(_, __) {
            throw new accounts_1.SignTransactionNotSupportedBySmartAccount();
        },
        async signTypedData(typedData) {
            return (0, actions_1.signTypedData)(client, {
                account: viemSigner,
                ...typedData
            });
        }
    });
    return {
        ...account,
        supportedKernelVersions: kernelVersion,
        validatorType: "SECONDARY",
        address: validatorAddress,
        source: "ECDSAValidator",
        getIdentifier() {
            return validatorAddress;
        },
        async getEnableData() {
            return viemSigner.address;
        },
        async getNonceKey(_accountAddress, customNonceKey) {
            if (customNonceKey) {
                return customNonceKey;
            }
            return 0n;
        },
        async signUserOperation(userOperation) {
            const hash = (0, permissionless_1.getUserOperationHash)({
                userOperation: {
                    ...userOperation,
                    signature: "0x"
                },
                entryPoint: entryPointAddress,
                chainId: chainId
            });
            const signature = await (0, actions_1.signMessage)(client, {
                account: viemSigner,
                message: { raw: hash }
            });
            return signature;
        },
        async getDummySignature() {
            return "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c";
        },
        async isEnabled(_kernelAccountAddress, _selector) {
            return false;
        }
    };
}
exports.signerToEcdsaValidator = signerToEcdsaValidator;
