// @zerodev/multi-chain-validator/_cjs/ecdsa/ecdsaSignUserOps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ecdsaSignUserOps = void 0;
const merkletreejs_1 = require("merkletreejs");
const permissionless_1 = require("permissionless");
const viem_1 = require("viem");
async function ecdsaSignUserOps({ account, multiUserOps, entryPoint: entryPointAddress }) {
    const userOpHashes = multiUserOps.map((multiUserOp, _index) => {
        return (0, permissionless_1.getUserOperationHash)({
            userOperation: {
                ...multiUserOp.userOperation,
                signature: "0x"
            },
            entryPoint: entryPointAddress,
            chainId: multiUserOp.chainId
        });
    });
    const merkleTree = new merkletreejs_1.MerkleTree(userOpHashes, viem_1.keccak256, {
        sortPairs: true
    });
    const merkleRoot = merkleTree.getHexRoot();
    const ecdsaSig = await account.kernelPluginManager.signMessage({
        message: {
            raw: merkleRoot
        }
    });
    const merkleProofs = userOpHashes.map((hash) => {
        return merkleTree.getHexProof(hash);
    });
    const multiChainSigs = multiUserOps.map((_, index) => {
        const merkleProof = merkleProofs[index];
        const encodedMerkleProof = (0, viem_1.encodeAbiParameters)([{ name: "proof", type: "bytes32[]" }], [merkleProof]);
        return (0, viem_1.concatHex)([ecdsaSig, merkleRoot, encodedMerkleProof]);
    });
    const signedMultiUserOps = multiUserOps.map((multiUserOp, index) => {
        return {
            ...multiUserOp.userOperation,
            signature: multiChainSigs[index]
        };
    });
    return signedMultiUserOps;
}
exports.ecdsaSignUserOps = ecdsaSignUserOps;
