// @zerodev/multi-chain-validator/_cjs/ecdsa/toMultiChainECDSAValidator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toMultiChainECDSAValidator = void 0;
const permissionless_1 = require("permissionless");
const accounts_1 = require("permissionless/accounts");
const accounts_2 = require("viem/accounts");
const actions_1 = require("viem/actions");
const actions_2 = require("viem/actions");
const constants_js_1 = require("../constants.js");
async function toMultiChainECDSAValidator(client, { signer, entryPoint: entryPointAddress, kernelVersion: _, validatorAddress }) {
    validatorAddress = validatorAddress ?? constants_js_1.MULTI_CHAIN_ECDSA_VALIDATOR_ADDRESS;
    const viemSigner = {
        ...signer,
        signTransaction: (_, __) => {
            throw new accounts_1.SignTransactionNotSupportedBySmartAccount();
        }
    };
    const chainId = await (0, actions_2.getChainId)(client);
    const account = (0, accounts_2.toAccount)({
        address: viemSigner.address,
        async signMessage({ message }) {
            return (0, actions_1.signMessage)(client, { account: viemSigner, message });
        },
        async signTransaction(_, __) {
            throw new accounts_1.SignTransactionNotSupportedBySmartAccount();
        },
        async signTypedData(typedData) {
            return (0, actions_1.signTypedData)(client, {
                account: viemSigner,
                ...typedData
            });
        }
    });
    return {
        ...account,
        supportedKernelVersions: ">=0.3.0",
        validatorType: "SECONDARY",
        address: validatorAddress,
        source: "MultiChainECDSAValidator",
        getIdentifier() {
            return validatorAddress ?? constants_js_1.MULTI_CHAIN_ECDSA_VALIDATOR_ADDRESS;
        },
        async getEnableData() {
            return viemSigner.address;
        },
        async getNonceKey(_accountAddress, customNonceKey) {
            if (customNonceKey) {
                return customNonceKey;
            }
            return 0n;
        },
        async signUserOperation(userOperation) {
            const hash = (0, permissionless_1.getUserOperationHash)({
                userOperation: {
                    ...userOperation,
                    signature: "0x"
                },
                entryPoint: entryPointAddress,
                chainId: chainId
            });
            const signature = await (0, actions_1.signMessage)(client, {
                account: viemSigner,
                message: { raw: hash }
            });
            return signature;
        },
        async getDummySignature(_userOperation) {
            return "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c";
        },
        async isEnabled(_kernelAccountAddress, _selector) {
            return false;
        }
    };
}
exports.toMultiChainECDSAValidator = toMultiChainECDSAValidator;
